File: stone_processor/src/main/java/com/github/klee0kai/stone/helpers/itemholder/SingleItemHolderHelper.java
package com.github.klee0kai.stone.helpers.itemholder;

import com.github.klee0kai.stone._hidden_.types.holders.SingleItemHolder;
import com.github.klee0kai.stone._hidden_.types.holders.StoneRefType;
import com.github.klee0kai.stone.helpers.codebuilder.SmartCode;
import com.squareup.javapoet.*;

import javax.lang.model.element.Modifier;

public class SingleItemHolderHelper implements ItemHolderCodeHelper {

    public String fieldName;
    public TypeName nonWrappedType;
    public TypeName returnType;

    public StoneRefType defRefType;
    public boolean isListCaching;


    @Override
    public FieldSpec.Builder cachedField() {
        ParameterizedTypeName cacheType = ParameterizedTypeName.get(ClassName.get(SingleItemHolder.class), nonWrappedType);
        return FieldSpec.builder(cacheType, fieldName, Modifier.PRIVATE, Modifier.FINAL)
                .initializer("new $T($T.$L)", cacheType, StoneRefType.class, defRefType.toString());
    }

    @Override
    public CodeBlock clearNullsStatement() {
        return CodeBlock.of("");
    }

    @Override
    public CodeBlock codeGetCachedValue() {
        String getMethod = isListCaching ? "getList" : "get";
        return CodeBlock.of("$L.$L()", fieldName, getMethod);
    }

    @Override
    public CodeBlock codeSetCachedValue(CodeBlock value, boolean onlyIfNull) {
        String setMethod = isListCaching ? "setList" : "set";
        return SmartCode.builder()
                .add(CodeBlock.of("$L.$L( ()-> ", fieldName, setMethod))
                .add(value)
                .add(CodeBlock.of(", $L)", onlyIfNull))
                .build(null);
    }

    @Override
    public CodeBlock statementSwitchRef(CodeBlock paramsCode) {
        return CodeBlock.builder()
                .addStatement("$L.switchCache($L)", fieldName, paramsCode)
                .build();
    }


}

File: stone_ksp/src/commonMain/kotlin/com/github/klee0kai/thekey/stone/ksp/helpers/itemholder/SingleItemHolderHelper.kt
package com.github.klee0kai.thekey.stone.ksp.helpers.itemholder

import com.github.klee0kai.stone.__hidden__.types.holders.SingleItemHolder
import com.github.klee0kai.stone.__hidden__.types.holders.StoneRefType
import com.github.klee0kai.thekey.stone.ksp.poet.genProperty
import com.github.klee0kai.thekey.stone.ksp.poet.smartcode.SmartCode
import com.github.klee0kai.thekey.stone.ksp.poet.smartcode.add
import com.github.klee0kai.thekey.stone.ksp.poet.smartcode.smartCode
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.KModifier
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.TypeSpec
import com.squareup.kotlinpoet.asClassName
import com.squareup.kotlinpoet.ksp.toClassName

class SingleItemHolderHelper(
    val fieldName: String,
    val returnType: KSType,
    val nonWrappedReturnType: KSType,
    val itemCacheType: ItemCacheType,
    val isListCaching: Boolean,
    val defRefType: StoneRefType,
) : ItemHolderHelper {

    override fun TypeSpec.Builder.genCacheField() {
        val cacheType = SingleItemHolder::class.asClassName()
            .parameterizedBy(nonWrappedReturnType.toClassName())

        genProperty(fieldName, cacheType) {
            addModifiers(KModifier.PRIVATE)
            initializer("%T(%T.%L)", cacheType, StoneRefType::class, defRefType)
        }
    }

    override fun codeGetCachedValue(
    ): SmartCode = smartCode {
        val getMethod = if (isListCaching) "getList" else "get"
        add("%L.%L()", fieldName, getMethod)
        providingType.value = returnType.toClassName()
    }

    override fun codeSetCachedValue(
        value: CodeBlock,
        onlyIfNull: Boolean
    ): CodeBlock = smartCode {
        val setMethod = if (isListCaching) "setList" else "set"
        add("%L.%L(onlyIfNull = %L ){ ", fieldName, setMethod, onlyIfNull)
        add(value)
        add("}")
        providingType.value = returnType.toClassName()
    }.collect()

}